packages/lodestar/src/sync/backfill/backfill.ts

```diff
@@ -0,0 +1,220 @@
+import {EventEmitter} from "events";
+import PeerId from "peer-id";
+import {StrictEventEmitter} from "strict-event-emitter-types";
+import {computeStartSlotAtEpoch} from "@chainsafe/lodestar-beacon-state-transition";
+import {IBeaconConfig} from "@chainsafe/lodestar-config";
+import {GENESIS_EPOCH} from "@chainsafe/lodestar-params";
+import {phase0, Root, Slot, allForks} from "@chainsafe/lodestar-types";
+import {ErrorAborted, ILogger} from "@chainsafe/lodestar-utils";
+import {List, toHexString} from "@chainsafe/ssz";
+import {IBeaconChain} from "../../chain";
+import {getMinEpochForBlockRequests} from "../../constants";
+import {IBeaconDb} from "../../db";
+import {INetwork, NetworkEvent, PeerAction} from "../../network";
+import {ItTrigger} from "../../util/itTrigger";
+import {PeerSet} from "../../util/peerMap";
+import {shuffleOne} from "../../util/shuffle";
+import {BackfillSyncError, BackfillSyncErrorCode} from "./errors";
+import {verifyBlockProposerSignature, verifyBlockSequence} from "./verify";
+
+/** Default batch size. Same as range sync (2 epochs) */
+const BATCH_SIZE = 64;
+
+export type BackfillSyncModules = {
+  chain: IBeaconChain;
+  db: IBeaconDb;
+  network: INetwork;
+  config: IBeaconConfig;
+  logger: ILogger;
+};
+
+export type BackfillSyncOpts = {
+  batchSize: number;
+};
+
+export enum BackfillSyncEvent {
+  completed = "BackfillSync-completed",
+}
+
+type BackfillSyncEvents = {
+  // slot param is oldest slot synced
+  [BackfillSyncEvent.completed]: (oldestSlotSynced: Slot) => void;
+};
+
+type BackfillSyncEmitter = StrictEventEmitter<EventEmitter, BackfillSyncEvents>;
+
+export class BackfillSync extends (EventEmitter as {new (): BackfillSyncEmitter}) {
+  private readonly chain: IBeaconChain;
+  private readonly network: INetwork;
+  private readonly db: IBeaconDb;
+  private readonly config: IBeaconConfig;
+  private readonly logger: ILogger;
+  private opts: BackfillSyncOpts;
+  /** last trusted block */
+  private anchorBlock: allForks.SignedBeaconBlock | null = null;
+  private processor = new ItTrigger();
+  private peers = new PeerSet();
+
+  constructor(modules: BackfillSyncModules, opts?: BackfillSyncOpts) {
+    super();
+    this.chain = modules.chain;
+    this.network = modules.network;
+    this.db = modules.db;
+    this.config = modules.config;
+    this.logger = modules.logger;
+    this.opts = opts ?? {batchSize: BATCH_SIZE};
+    this.network.events.on(NetworkEvent.peerConnected, this.addPeer);
+    this.network.events.on(NetworkEvent.peerDisconnected, this.removePeer);
+
+    this.sync()
```

Comment:
Merged sync() and init() functions to it's easy to detect a completed event in all cases, even when early exit conditions are found